src/vivarium_census_prl_synth_pop/results_processing/addresses.py

```diff
@@ -18,6 +19,98 @@
 }
 
 
+def get_address_id_maps(
+    column_name: str,
+    obs_data: Dict[str, pd.DataFrame],
+    artifact: Artifact,
+    randomness: RandomnessStream,
+):
+    """
+    Get all maps that are indexed by `address_id`.
+
+    Parameters
+    ----------
+    column_name
+        Name of the column to use as an index
+    obs_data
+        Observer DataFrame with key for the observer name
+    artifact
+        A vivarium Artifact object needed by mapper
+    randomness
+        RandomnessStream to use in choosing zipcodes proportionally
+
+    Returns
+    -------
+    A dictionary of pd.Series suitable for pd.Series.map, indexed by `address_id`
+
+    """
+    if column_name != "address_id":
+        raise ValueError(f"Expected `address_id`, got `{column_name}`")
+    maps = dict()
+    maps.update(get_zip_map(column_name, obs_data, randomness))
+    maps.update(get_household_address_map(column_name, obs_data, artifact, randomness))
+    return maps
+
+
+def get_zip_map(
+    column_name: str,
+    obs_data: Dict[str, pd.DataFrame],
+    randomness: RandomnessStream,
+) -> Dict[str, pd.Series]:
+    """Gets a mapper for `address_id` to zipcode, based on state, puma, and proportion.
+
+    Parameters
+    ----------
+    column_name
+        Name of the column to use as an index
+    obs_data
+        Observer DataFrame with key for the observer name
+    randomness
+        RandomnessStream to use in choosing zipcodes proportionally
+
+    Returns
+    -------
+    A pd.Series suitable for pd.Series.map, indexed by column_name, with key "zipcode"
+
+    """
+    zip_map_dict = {}
+    output_cols = [column_name, "state", "puma"]  # columns in the output we use to map
+    address_ids = format_data_for_mapping(
+        index_name=column_name,
+        obs_results=obs_data,
+        output_columns=output_cols,
+    )
+    zip_map = pd.Series(index=address_ids.index)
+
+    # Read in CSV and normalize
+    map_data = pd.read_csv(PUMA_TO_ZIP_DATA_PATH)
+    proportions = (
+        map_data.groupby(["state", "puma"])
+        .apply(sum)["proportion"]
+        .reset_index()
+        .set_index(["state", "puma"])
+    )
+    normalized_groupby = (
+        (map_data.set_index(["state", "puma", "zipcode"]) / proportions)
+        .reset_index()
+        .groupby(["state", "puma"])
+    )
+
+    for (state, puma), df_locale in address_ids.groupby(["state", "puma"]):
```

Comment:
We can proceed with this for now, but this is a python loop over a large number of rows - there are a lot of state/puma combinations. I haven't looked at this long enough to have a solution off the top of my head, but I'd be surprised if there wasn't a vectorized operation we could do here rather than looping over the groupby in python code.

Reply:
I'll run this against real data and see how bad the inefficiency is and log to the profiling page.